File: stream.go
package unio

import (
	"context"
	"errors"

	"github.com/Fullstop000/unio/driver"
)

// Stream exposes the observable events of one Run.
type Stream struct {
	ctx     context.Context
	ctxDone <-chan struct{}
	owner   *Session
	events  <-chan driver.AgentEvent
	runID   driver.RunID

	cur       Event
	done      bool
	result    Result
	err       error
	text      []byte
	thinking  []byte
	cancelErr error
}

func newStream(ctx context.Context, owner *Session, events <-chan driver.AgentEvent, runID driver.RunID) *Stream {
	return &Stream{ctx: ctx, ctxDone: ctx.Done(), owner: owner, events: events, runID: runID}
}

// Next advances to the next text, thinking, tool-call, or tool-result event.
func (s *Stream) Next() bool {
	if s.done {
		return false
	}
	for {
		select {
		case ev, ok := <-s.events:
			if !ok {
				if err := s.ctx.Err(); err != nil {
					s.finish(Result{}, err, Idle)
					return false
				}
				s.finish(Result{}, driver.NewTransportError("unio: event stream closed before completion"), Idle)
				return false
			}
			if ev.Type == driver.EventSessionAttached {
				if err := s.owner.setID(ev.SessionID); err != nil {
					s.finish(Result{}, err, Idle)
					return false
				}
				continue
			}
			if ev.RunID != s.runID {
				continue
			}
			switch ev.Type {
			case driver.EventOutput:
				if ev.Item.Kind == driver.ItemTurnEnd {
					continue
				}
				s.cur = s.accumulate(ev.Item)
				return true
			case driver.EventCompleted:
				s.result.Text = string(s.text)
				s.result.Thinking = string(s.thinking)
				s.result.SessionID = ev.SessionID
				s.result.Usage = ev.Result.Usage
				s.result.DurationMs = ev.Result.DurationMs
				s.result.Interrupted = ev.Result.FinishReason == driver.FinishCancelled
				_ = s.owner.setID(ev.SessionID)
				if ev.Result.FinishReason == driver.FinishTransportClosed {
					s.owner.dropAttachment()
					s.finish(s.result, driver.NewTransportError("agent transport closed during turn"), Idle)
					return false
				}
				s.finish(s.result, s.cancelErr, Idle)
				return false
			case driver.EventBlocked:
				s.result.Text = string(s.text)
				s.result.Thinking = string(s.thinking)
				s.result.SessionID = ev.SessionID
				if ev.Blocked != nil {
					reason := *ev.Blocked
					s.result.Blocked = &reason
				}
				s.finish(s.result, nil, Blocked)
				return false
			case driver.EventFailed:
				s.result.Text = string(s.text)
				s.result.Thinking = string(s.thinking)
				s.finish(s.result, failedToError(ev), Idle)
				return false
			}
		case <-s.ctxDone:
			s.cancelErr = s.ctx.Err()
			s.ctxDone = nil
			err := s.owner.Interrupt()
			if err != nil {
				s.finish(s.result, errors.Join(s.cancelErr, err), Idle)
				return false
			}
		}
	}
}

// Event returns the event produced by the latest successful Next call.
func (s *Stream) Event() Event { return s.cur }

// Result drains the stream and returns its accumulated result.
func (s *Stream) Result() (Result, error) {
	for s.Next() {
	}
	return s.result, s.err
}

func (s *Stream) accumulate(item driver.AgentEventItem) Event {
	switch item.Kind {
	case driver.ItemText:
		s.text = append(s.text, item.Text...)
	case driver.ItemThinking:
		s.thinking = append(s.thinking, item.Text...)
	case driver.ItemToolCall:
		s.result.ToolCalls = append(s.result.ToolCalls, ToolCall{Name: item.Tool, Input: item.ToolInput})
	}
	return Event{Kind: EventKind(item.Kind), Text: item.Text, Tool: item.Tool, ToolInput: item.ToolInput}
}

func (s *Stream) finish(result Result, err error, state SessionState) {
	if s.done {
		return
	}
	s.done = true
	s.result = result
	s.err = err
	s.owner.setState(state)
}

File: python/src/unio/_stream.py
from __future__ import annotations

from typing import TYPE_CHECKING

from ._driver import DriverEventType, EventQueue, FinishReason, OutputKind
from .errors import runtime_reported, transport
from .models import Event, EventKind, Result, SessionState, ToolCall

if TYPE_CHECKING:
    from ._session import Session


class Stream:
    """The async event stream and accumulated result of one turn."""

    def __init__(self, owner: Session, events: EventQueue, run_id: str) -> None:
        self._owner = owner
        self._events = events
        self._run_id = run_id
        self._result = Result()
        self._done = False
        self._error: BaseException | None = None

    def __aiter__(self) -> Stream:
        return self

    async def __anext__(self) -> Event:
        if self._done:
            raise StopAsyncIteration
        try:
            while True:
                event = await self._events.get()
                if event is None:
                    self._finish(error=transport("event stream closed before completion"))
                    raise StopAsyncIteration
                if event.type is DriverEventType.SESSION_ATTACHED:
                    self._owner._set_id(event.session_id)
                    continue
                if event.run_id != self._run_id:
                    continue
                if event.type is DriverEventType.OUTPUT and event.item is not None:
                    if event.item.kind is OutputKind.TURN_END:
                        continue
                    return self._accumulate(event.item)
                if event.type is DriverEventType.COMPLETED:
                    if event.result is not None:
                        self._result.usage = dict(event.result.usage)
                        self._result.duration_ms = event.result.duration_ms
                        self._result.interrupted = (
                            event.result.finish_reason is FinishReason.CANCELLED
                        )
                        if event.result.finish_reason is FinishReason.TRANSPORT_CLOSED:
                            self._finish(error=transport("agent transport closed during turn"))
                            raise StopAsyncIteration
                    self._result.session_id = event.session_id
                    self._owner._set_id(event.session_id)
                    self._finish()
                    raise StopAsyncIteration
                if event.type is DriverEventType.BLOCKED:
                    self._result.session_id = event.session_id
                    self._result.blocked = event.blocked
                    self._finish(state=SessionState.BLOCKED)
                    raise StopAsyncIteration
                if event.type is DriverEventType.FAILED:
                    self._finish(error=event.error or runtime_reported("agent run failed"))
                    raise StopAsyncIteration
        except BaseException as error:
            if not isinstance(error, (StopAsyncIteration, GeneratorExit)) and not self._done:
                await self._owner.interrupt()
                self._finish(error=error)
            raise

    async def result(self) -> Result:
        """Drain remaining events and return the accumulated terminal Result."""
        async for _ in self:
            pass
        if self._error is not None:
            raise self._error
        return self._result

    def _accumulate(self, item: object) -> Event:
        from ._driver import OutputItem

        assert isinstance(item, OutputItem)
        kind = EventKind(item.kind.value)
        if kind is EventKind.TEXT:
            self._result.text += item.text
        elif kind is EventKind.THINKING:
            self._result.thinking += item.text
        elif kind is EventKind.TOOL_CALL:
            self._result.tool_calls.append(ToolCall(item.tool, item.tool_input))
        return Event(kind=kind, text=item.text, tool=item.tool, tool_input=item.tool_input)

    def _finish(
        self,
        *,
        state: SessionState = SessionState.IDLE,
        error: BaseException | None = None,
    ) -> None:
        if self._done:
            return
        self._done = True
        self._error = error
        self._owner._finish_turn(state)
